plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/internal/Version.java

```diff
@@ -0,0 +1,374 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.bolyuba.nexus.plugin.npm.internal;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Version following semantic defined by <a href="http://semver.org/">Semantic Versioning</a> document.
+ *
+ * Copied from https://github.com/jeluard/semantic-versioning/blob/master/api/src/main/java/org/semver/Version.java
+ * Retained original header as ASL2 requires. Commit 95232052f900ac565917246b29b56b058cc060da.
+ */
+@Immutable
+public final class Version implements Comparable<Version> {
+
+    /**
+     * {@link Version} element. From most meaningful to less meaningful.
+     */
+    public enum Element {
+        MAJOR, MINOR, PATCH, SPECIAL;
+    }
+
+    private static final String FORMAT = "(\\d+)\\.(\\d+)(?:\\.)?(\\d*)(\\.|-|\\+)?([0-9A-Za-z-.]*)?";
+    private static final Pattern PATTERN = Pattern.compile(Version.FORMAT);
+
+    private static final Pattern DIGITS_ONLY = Pattern.compile("\\d+");
+
+    private static final String SNAPSHOT_VERSION_SUFFIX = "SNAPSHOT";
+
+    private final int major;
+    private final int minor;
+    private final int patch;
+    private final String separator;
+    private final Special special;
+
+    public Version(@Nonnegative final int major, @Nonnegative final int minor, @Nonnegative final int patch) {
+        this(major, minor, patch, null, null);
+    }
+
+    public Version(@Nonnegative final int major, @Nonnegative final int minor, @Nonnegative final int patch, @Nullable final String separator, @Nullable final String special) {
+        if (major < 0) {
+            throw new IllegalArgumentException(Element.MAJOR+" must be positive");
+        }
+        if (minor < 0) {
+            throw new IllegalArgumentException(Element.MINOR+" must be positive");
+        }
+        if (patch < 0) {
+            throw new IllegalArgumentException(Element.PATCH+" must be positive");
+        }
+
+        this.major = major;
+        this.minor = minor;
+        this.patch = patch;
+        this.separator = separator;
+        this.special = parseSpecial(special);
+    }
+
+    private Special parseSpecial(String specialString) {
+      if (specialString == null) {
+        return null;
+      }
+      Special special = new Special(specialString);
+      if (special.ids.length == 0) {
+        return null;
+      }
+      return special;
+    }
+
+    /**
+     *
+     * Creates a Version from a string representation. Must match Version#FORMAT.
+     *
+     * @param version
+     * @return
+     */
+    public static Version parse(@Nonnull final String version) {
+        final Matcher matcher = Version.PATTERN.matcher(version);
+        if (!matcher.matches()) {
```

Comment:
@mrprescott this is C&P from https://github.com/jeluard/semantic-versioning/blob/bca36dc2202f806ecc022b8b34f110eba830fe9e/api/src/main/java/org/semver/Version.java